motionplan/motionPlanner.go

```diff
@@ -457,30 +457,30 @@ func CheckPlan(
 	currentInputs []frame.Input,
 	errorState spatialmath.Pose,
 	logger *zap.SugaredLogger,
-) error {
+) (spatialmath.Pose, error) {
```

Comment:
Why is this a return type? What is the significance of this variable

Reply:
This returns first interpolated pose which results in a collision. Comment added both to this function and to return describing this.

Reply:
With the interpolated position being returned, would we like to change the CheckPlan function to return the collision pose and a nil err in the event of a collision? Previously an error would be returned with collision information instead of it.

Reply:
`CheckPlan` to me communicates a boolean result (in this case the error being nil or no is effectively this).  Now we're returning a pose which is a somewhat surprising variable for this name.  We should either change the name of this function to encapsulate the new behavior or make a separate function that reuses the same code

Reply:
It makes sense to me to have some lookahead functionality; if we detect a transient obstacle 500m ahead, we probably shouldn't care until we are closer.

I have less of an opinion regarding whether we only look ahead that far, or check the whole plan and return the pose of the closest collision. Though the more I think about it, I lean more towards the former.

Reply:
I agree that the former with me more computationally efficient. However, it will be a great deal more complicated to implement if we want to generalize it to PTGs. I'd prefer to hold off on that in this PR unless the motion team feels strongly about handling this now

Reply:
@nfranczak copy and pasting a comment I made that relates to this thread: 

I potentially disagree with this assertion because:

Returning a position of obstacle collision is useful information for a variety of use cases, more so than just a bool which can be gotten straight from the error being nil or not
We may want to allow users to configure their collision distance in the future and so would have to pass this distance value into it.
Passing in this value takes away from the CheckPlan nature and would make this function more accurately performing a CheckPartialPlan.
Explore mode does not use waypoints just interpolated start and end points to segments
I may be overlooking a key feature of waypoints however, if so please let me know. I’m just more in favorite of letting CheckPlan return the most useful and ubiquitous information (i.e. position of collision) and letting the explore motion service use that information to handle when stop said move action (i.e. an obstacle within 10 cm)

Reply:
> However, it will be a great deal more complicated to implement if we want to generalize it to PTGs

Disagree; this could be as simple as adding a check inside the `for _, interpConfig := range interpolatedConfigurations {` loop saying something like
```
if poseInPath.Point().Norm() > lookaheadDistance {
   return nil, nil
}
```
That will work robustly for PTGs as well, and also deals with the case of a path doubling back on itself.

>  Passing in this value takes away from the CheckPlan nature and would make this function more accurately performing a CheckPartialPlan.

I'm fine with renaming to CheckPartialPlan.

Reply:
> I'm fine with renaming to CheckPartialPlan.

I am also fine with renaming to `CheckPartialPlan`.
I also like the idea of adding a new input to the function which checks if we are past `lookaheadDistanceMM`.
If we set `lookaheadDistanceMM` high enough we can still check the entire plan.

Reply:
> Returning a position of obstacle collision is useful information for a variety of use cases, more so than just a bool which can be gotten straight from the error being nil or not.

I am unsure how useful returning the position of a collision is.
The user wouldn't be interacting this information in real time and would have to grep the logs to find this information to deliver some analysis. Further, the position of the collision is difficult to understand for this specific use case. The robot doesn't know where it is at all times so returning a collision relative to an object which has an indeterminate position seems difficult to reason about...

We wouldn't return a `bool` if we are not returning the collision of a pose, rather just an error which may or may not be nil.

Reply:
I like the idea of specifying a lookahead distance a lot.  This would mean that the only return needed would be the error we are returning today, correct?

Reply:
I'm honestly a little worried about CheckPlan.

The design we have now assumes that only the remaining pieces of the plan is passed in; which is not what is happening in this PR. This makes it very hard to tell where we actually are in the plan.

We should probably enforce that "plan must not include fully-executed waypoints"